### PR TITLE
User register and support permissions on UI

### PR DESCRIPTION
**Pull request by krisx303**
Changes:
* Endpoint for registering new user (called automatically right after registering on UI)
* Endpoint for searching users with applied filter (email or display name)
* Endpoint for retrieving full permission model to resource with detailed users which have permissions granted
* Fixed logic PermissionAccessService
* Support for sysadmin claim set on firebase (PreAuthorize annotations)

### CODE REVIEW

File: src/main/java/com/brightpath/learnify/config/DevDatabaseConfigurator.java [hunk 1]
```diff
@@ -16,12 +16,11 @@
 import org.springframework.boot.CommandLineRunner;
 import org.springframework.context.annotation.Bean;
 import org.springframework.context.annotation.Configuration;
-import org.springframework.context.annotation.Profile;
 
 import java.util.List;
 
 @Configuration
-@Profile("dev")
+//@Profile("dev")
```
Comment (krisx303) [inline src/main/java/com/brightpath/learnify/config/DevDatabaseConfigurator.java]: revert this

Comment (Adam3004) [reply src/main/java/com/brightpath/learnify/config/DevDatabaseConfigurator.java]: surely

File: src/main/java/com/brightpath/learnify/config/FirebaseDatabaseConfigurator.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+@RequiredArgsConstructor
+public class FirebaseDatabaseConfigurator {
+
+    private final UserRepository userRepository;
+
+    @Bean
+    CommandLineRunner firebaseUsersSync() {
+        return args -> {
+            try {
+                ListUsersPage page = FirebaseAuth.getInstance().listUsers(null);
```
Comment (Adam3004) [inline src/main/java/com/brightpath/learnify/config/FirebaseDatabaseConfigurator.java:26]: won't page be always empty if we pass null as argument in listUsers?

File: src/main/java/com/brightpath/learnify/controller/BindingController.java [hunk 3]
```diff
@@ -13,41 +12,47 @@
...
     public ResponseEntity<List<NoteSummaryDto>> listNotesBoundToQuiz(UUID quizId) {
-        permissionAccessService.checkUserPermissionToViewResource(quizId, QUIZ);
         List<Note> notes = bindingService.listNotesBoundToQuiz(quizId);
         return ResponseEntity.ok(notes.stream().map(dtoMapper::asNoteSummaryDto).toList());
     }
 
     @Override
+    @PreAuthorize("""
+                    @permissionAccessService.checkUserPermissionToViewResource(#noteId, 'NOTE') or
+                    @userIdentityService.isCurrentUserAdmin()
```
Comment (Adam3004) [inline src/main/java/com/brightpath/learnify/controller/BindingController.java:53]: i really like this option

File: src/main/java/com/brightpath/learnify/domain/auth/PermissionAccessService.java [hunk 4]
```diff
@@ -37,11 +40,30 @@ public class PermissionAccessService {
...
+                userPermissions
+        );
+    }
 
 
     // method for checking if user has access to a resource with a given id and type
     public ResourceAccessEnum getUserAccessForResource(String userId, UUID resourceId, ResourceType resourceType) {
         ResourceAccessSummary access = permissionAccessRepository.findUserAccessToResource(userId, permissionAccessId(resourceId, resourceType));
+        if (access == null) {
+            throw new ResourceNotFoundException(resourceType);
```
Comment (Adam3004) [inline src/main/java/com/brightpath/learnify/domain/auth/PermissionAccessService.java:65]: is it right exception for the situation? No access found doesn't mean that the resource doesn't exists. But we can leave it as it is

File: src/main/java/com/brightpath/learnify/domain/auth/PermissionAccessService.java [hunk 5]
```diff
@@ -82,80 +104,63 @@ public List<Permission> getPermissionsForResource(UUID resourceId, ResourceType
...
-            throw new UserNotAuthorizedToGetException();
-        }
+        PermissionsAccessEntity permissionsAccessEntity = permissionAccessRepository.findById(permissionAccessId(resourceId, resourceType))
+                .orElseThrow(() -> new ResourceNotFoundException(resourceType));
+        return permissionsAccessEntity.getOwnerId().equals(userId);
     }
 
     public Permission addPermissionToResourceForUser(UUID resourceId, ResourceType resourceType, String userId, ResourceAccessEnum requestedAccess) {
-        if (userHasAnyPermissionToResource(resourceId, resourceType, userId)) {
+        userService.checkIfUserExists(userId);
```
Comment (Adam3004) [inline src/main/java/com/brightpath/learnify/domain/auth/PermissionAccessService.java:144]: nice idea

**Review by Adam3004 — APPROVED** [review]

File: src/main/java/com/brightpath/learnify/config/FirebaseDatabaseConfigurator.java (hunk 2 above)
Comment (krisx303) [reply src/main/java/com/brightpath/learnify/config/FirebaseDatabaseConfigurator.java:26]: It returns the first page of users:
_Gets a page of users starting from the specified pageToken. Page size is limited to 1000 users.
Params:
pageToken – A non-empty page token string, or null to retrieve the first page of users.
Returns:
A ListUsersPage instance._

File: src/main/java/com/brightpath/learnify/controller/BindingController.java (hunk 3 above)
Comment (krisx303) [reply src/main/java/com/brightpath/learnify/controller/BindingController.java:53]: nice to hear that

File: src/main/java/com/brightpath/learnify/domain/auth/PermissionAccessService.java (hunk 4 above)
Comment (krisx303) [reply src/main/java/com/brightpath/learnify/domain/auth/PermissionAccessService.java:65]: In the SQL query there is `LEFT JOIN` on user permissions, so there are few cases: 
* If there is permission record for requested user - the object is returned with all fields present (specified to single user)
* If there is no permission record for requested user - the object is returned only with basic fields (global permission level and ownerID)
* If there is no permission record AND there is no record in PermissionAccess the null is returned - I assumed it is impossible to have resource without associated permission model

**Pull request merged**